Reorder tables with single SQL query (all DBs compatible), greatly improving reorder performance

ggppdk opened the pull request:
**Performance fix** for JTable::reorder(),
- fixes performance issue on new article save: #10567
- ~~greatly improves performance / race condition issues on drag and drop reordering and break-up: #4303~~

It should work in all Database servers, since it uses **case/when/then/else/end expression**,  (SQL-92)
#### Summary of Changes
- Replace multiple reordering SQL queries with a single query
- Thus updates 10,000 records per query during re-ordering
#### Testing Instructions

1 - Applying PR to **latest staging**
2 - It helps to make the per row ORDER inputs visible, to do it add CSS to your backend template file:   **/templates/isis/css/template.css**

```
.text-area-order, .sortable-handler + input {
  display: inline !important;  width: 40px; text-align: right;
}
```

You may want to add just before JTable::reorder() final return;
https://github.com/joomla/joomla-cms/pull/11184/files#diff-8254e2c441a41d3fa26f0f83fbbb3043R1431

```
JFactory::getApplication()->enqueueMessage('<b>reorder()</b>:<br/>'.$query, 'message');
```
1. Then create new article 
2. Click "Save and close" (not "Save", since we want to see the enqueued message and also list the orderings in article manager) see an example in picture

![table_reorder_single_query](https://cloud.githubusercontent.com/assets/5092940/16932698/c9703da8-4d50-11e6-9d12-f0e983f62494.png)

ggppdk committed:
Reorder table with single SQL query

bertmert commented on libraries/joomla/table/table.php:
```diff
@@ -1365,6 +1365,17 @@ public function reorder($where = '')
 		$this->_db->setQuery($query);
 		$rows = $this->_db->loadObjectList();
+		// Reorder configuration
+		$uses_single_key = count($this->_tbl_keys == 1);
```
Shouldn't this be:
`$uses_single_key = count($this->_tbl_keys) == 1;`

ggppdk committed:
Fixed wrong count

csthomas commented:
If joomla users want to stay with old ordering then 
what about use a transaction start before loop and commit after it.
This also speed up queries.
Joomla use innoDB, postresql and sqlazure which support database transactions.

Your patch will do the same but it is more complicated.

ggppdk commented:
What is complicated about it ?

it is actually very simple collect all updates in 1 query

also a transaction with multiple queries should always be slower or a lot slower,
or do you mean transaction with 1 query ?

bertmert commented:
I have tested this item :red_circle: unsuccessfully on c43254e00226d501f7bfc255038df41a047edc3a<br /><br />I would prefer this solution against https://github.com/joomla/joomla-cms/pull/11139 because ordering numbers are more transparent. Starting at 1 or so instead of 2147483647.

(But the more I think about this whole reordering issue I don't understand why we have to reorder the table while saving. Ordering is an unsigned field. Why not just calculating min_ordering once and set ordering of new article(s) to (--min_ordering). If not possible because of field size, only then do a reorder starting at 1.)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11184">issues.joomla.org/joomla-cms/11184</a>.</sub>

bertmert commented:
I have tested this item :white_check_mark: successfully on c43254e00226d501f7bfc255038df41a047edc3a<br /><br />I would prefer this solution against https://github.com/joomla/joomla-cms/pull/11139 because ordering numbers are more transparent. Starting at 1 or so instead of 2147483647.

(But the more I think about this whole reordering issue I don't understand why we have to reorder the table while saving. Ordering is an unsigned field. Why not just calculating min_ordering once and set ordering of new article(s) to (--min_ordering). If not possible because of field size, only then do a reorder starting at 1.)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11184">issues.joomla.org/joomla-cms/11184</a>.</sub>

ggppdk committed:
Add new line for code readability

ggppdk commented:
Anyone testing, 
if you want to test drag and drop ordering, please be warned to remove the CSS change
```
body  .text-area-order {
  display: inline !important;
}
```
as it breaks drag and drop ordering (does the JS selector need the input elements to be hidden ?)

bertmert commented:
> if you want to test drag and drop ordering, please be warned to remove the CSS change

I didn't and had no issues. Fields don't update without page reload. That's all

Off Topic:
Nice thing (for my own private site) is that one can enter a number (e.g. -25). After drag&drop -25 is stored correctly in the database. You have to reload the page to see this ordering change in backend.

csthomas commented:
Please take a look at Review of solutions https://github.com/joomla/joomla-cms/pull/11139#issuecomment-233566830

ggppdk committed:
Minor move out of loop, no performance gain since loop is executed once or twice

ggppdk commented:
I think, i will update this PR after this PR is tested:  #11139

so that to make it include only the changes to JTable::reorder()
- so please don't test yet

ggppdk committed:
Removed changes from model

csthomas commented:
I'm back to there because this is more related.
To be less race conditions you would only add one transaction.

```sql
BEGIN

-- loop on each 10000 queries: 
--    UPDATE max 10000 rows
 
COMMIT
```

ggppdk commented:
this PR fixes performance issue and makes race issue very rare

but transactions are not be directly related to this PR

Joomla database API needs to add them first ?

- checking if current DB supports them
- checking if current sql user is privileged to do them
- adding some API calls for them

right ?

csthomas commented:
https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/database/driver.php#L2154

ggppdk commented:
So it is safe to use regardless of enironment ?

i will make use of it,
thanks, small change in code of this PR

csthomas commented:
I have tested `transactionStart` or on my own component.

You can see some example on fof library:
https://github.com/joomla/joomla-cms/blob/staging/libraries/fof/table/nested.php#L181

ggppdk commented:
@alikon 

I have managed pulled it locally and rebase it 
(originally i made it from the GitHub GUI)

So it is good to test when you finished your other work (as you said)
- especially **in non-Mysql DBs** and see if some modification is needed for them 

**case/when/then/else/end** expression,  was added by **SQL-92**

so it should be supported by all DBs ?

alikon commented:
added to my to-do list :+1: 
not so sure i will be able to test on MSSQL

p.s.
seems travis is not happy

ggppdk committed:
Fixed PHP code styling

ggppdk commented:
@alikon
Done, fixed the code styling,

Also updated the description of the first posting, to be up to date, shorter and clearer.
Any other testers welcome, especially if they can also test the non-mysql DBs too

mbabker commented:
I'd suggest checking your limits with the SQL Server docs.  For an INSERT query, it only allows 1000 values at a time (see https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_finder/helpers/indexer/driver/sqlsrv.php#L543 for Smart Search where it hit this issue).

ggppdk commented:
@mbabker 
that query tries to do the same thing for the case of inserts
- i see 1 query per 1000 inserts

To be honest in this case i don't know of what limitation applies to size of queries of SQL server

but, in our case the updates **are really small strings**
- e.g. for MySQL **max_packet_size default value 1MByte** , (some servers change it to much larger)

Let's say that the code below is 30 characters
WHEN 'NNNNN'  THEN OOOOOO

you will get 
1,000,000 / 30  =  33,000 updates (ok if you count some 10% overhead this maybe smaller)
to fit in the default MAX packet size of MySQL

in the code of this, i use in the 10,000 updates per query

i do not know if SQL server has smaller limits (i need to check the docs ...)

we can probably change it to 1,000 with small or little performance drop

If the values were mostly variable length,
we can also count the characters, (that is what i do in my queries to avoid hitting the limit)

ggppdk commented:
i will check the docs, but we also need some category with 30,000 articles for testing this properly

brianteeman commented:
I can give you that or you can make your own using com_overload

ggppdk commented:
Thanks brian, i have for MySql,

i have made a patch for com_overload to make it run faster, and had created a website with 500,000 articles

but i have not setup any Joomla web-sites in other DBs

mbabker commented:
I mainly just want to make sure UPDATE queries in SQL Server don't have the
same data limitations as INSERT queries, or others that are more
restrictive than the other engines.  We've got a couple dozen users to keep
happy ya know 😆

On Saturday, August 13, 2016, Georgios Papadakis <notifications@github.com>
wrote:

> i will check the docs, but we also need some category with 30,000 articles
> for testing this properly
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11184#issuecomment-239640866>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWfocbTYXWcJJB5UYvv7MlV1wQYqOe5ks5qfi3fgaJpZM4JPNjv>
> .
>

ggppdk commented:
I hope this PR is proven good to be accepted for the other DBs too,
it will make many 3rd party extensions happy too )))

brianteeman commented:
If you are on a mac then checkout http://postgresapp.com/ its super easy

On 13 August 2016 at 22:02, Georgios Papadakis <notifications@github.com>
wrote:

> I hope this PR is proven good to be accepted for the other DBs too,
> it will make many 3rd party extensions happy too )))
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11184#issuecomment-239641403>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPH8foLBEE0jZR5n0RJJBqcF42BX8reks5qfjDrgaJpZM4JPNjv>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

Sieger66 commented:
> i have made a patch for com_overload to make it run faster, 
@ggppdk 
Can you give us the patch  for com_overload to make it easier to test the PR #11184

for example create there https://github.com/nikosdion/com_overload/issues
a "New issue" and attach file/s

ggppdk commented:
aahh, I need to remake it, i deleted several test installations, and localhost copies of sites,
to clean my local drives, it was gone with the cleanup

andrepereiradasilva commented:
what's the state of this?

ggppdk commented:
Good for testing

andrepereiradasilva commented:
Humm didn't work for me i guess. please check animated gif
![reorder](https://cloud.githubusercontent.com/assets/9630530/18212903/923ec118-713d-11e6-9611-82668c331722.gif)

ggppdk commented:
I retested just now,  it works, 
- saving new article
- reordering

what the animated gif does not seem to be relevant to this PR
- also you are displaying the select query in you animated gif, but i had suggested to display **the UPDATE query**, so that we can see the actual update taking place (after you click save and close)

ggppdk commented:
Here how the update queries should be after after "save and close":
![reorder1](https://cloud.githubusercontent.com/assets/5092940/18215863/2ab86552-715c-11e6-8db6-20bae69db225.png)

andrepereiradasilva commented:
Ups sorry. Will retweet latter

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on cdc5359f1f39e0d40e47b50717b10d61b47c1873<br /><br />Tested and ordering seems fine.
Notice also a performance improvement: 400 articles in the same category
- Before patch: 82.4130 ms
- After patch: 16.5069 ms<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11184">issues.joomla.org/joomla-cms/11184</a>.</sub>

alikon commented:
i'm testing on postgresql with a category with 500 articles in one language

**testing creating a new article**

- before patch 1.818 seconds

- after patch 0.125 seconds

but i'm quite concerned about how long is the text of SQL query (see animated gif) and how log it can became with more content,

suppose something like 100k content, i'm afraid we can exceed the limits of some common basic hosting server configuration  in this way (especially cheap hosting with mysql basic configuration

![test11084](https://cloud.githubusercontent.com/assets/181681/18224160/1ad5864e-71cd-11e6-925d-49108b1525a8.gif)


even if the  #8576  has been reverted for B/C reason IMHO it is the most easy way to manage ordering
again for me is not correct to reorder all data  when creating  new  items

born2webdesign commented:
From what I gather, this looks like a nice interim solution …
But I agree with @alikon (and the problem-inducing change in 3.6) that reordering on inserting is a bad idea. However, I have been bitten by the 3.6 change, too.
Couldn't this be a way forward:
- Offer an option to insert new items at the end in the DB (and make this the default in new installations).
- Offer a solution (button ?) to reorder existing items (for the category being displayed or for everything?).
- Then, there should be a "Reverse Article Order" option for the frontend.
Probably with some post-installation message to inform/guide existing users.
This way, we could make the new behaviour the default for new installations; but existing installations wouldn't break (and get improved performance with #11184 ), and could make the switch at some point.
Sorry, I don't have any code to offer for this, but it seems like the "right thing"™ to me.

ggppdk commented:
> but i'm quite concerned about how long is the text of the SQL query (see animated gif) and how long it can became with more content

Actually there is a limit to the number of records, current limit is 10000, see:
https://github.com/joomla/joomla-cms/pull/11184/files#diff-8254e2c441a41d3fa26f0f83fbbb3043R1374

which will make a query of 10,000 x 30 characters (WHEN ... THEN ...) thus less than< 300 KBytes, and this much lower than the MySQL default max_packet_size which is 1MByte (all hosts have default or more)

See comment
https://github.com/joomla/joomla-cms/pull/11184#issuecomment-239640737

andrepereiradasilva commented:
so if we have 1.000.000 records we get 120 queries! to much queries! :smiling_imp:

csthomas commented:
@andrepereiradasilva If you have 1.000.000 featured articles then yes. 
On each added featured article. But it is better than 1.000.000 queries now.

andrepereiradasilva commented:
i was jooking :wink:

alikon commented:
> which will make a query of 10,000 x 30 characters (WHEN ... THEN ...) thus less than< 300 KBytes

plus for the `IN()` 

10,000 x 10 characters [ IN (nnnnnnnn1, nnnnnnnn2, ......., nnnnnnnn) less than < 100 KBytes

so still under default max_packet_size 1MB ( should be the same on postgres ) 

call to @waader  for a test on mssql

waader commented:
Reordering works with testing data installed but because of this #11932 i was not able to create new articles. So I could not follow the test instructions.

ggppdk commented:
Also during the time i was making this PR i tested by using limit of 5 or 50 values to confirm that breaking the update Query into more than 1 update queries works:

https://github.com/joomla/joomla-cms/pull/11184/files#diff-8254e2c441a41d3fa26f0f83fbbb3043R1374

so you may want to test by adding a 5 or 50 limit to the above (depending on how many articles / records your category has)

thus the update queries, for updating 500 records will be 10 UPDATE queries

ggppdk commented:
The change of using last order number in new record creation was reverted
so J3.6.3, will now again have same slow down on new records saving in site with many records per category

This PR does not have a milestone

DavorMtc commented:
After upodating my joomla to 3.6.3 (from 3.6.2) I had problems with "504 gateaway" error when saving new articles.

Found this PR, did everything like its written and now its working fine.

ggppdk commented:
Since PR: https://github.com/joomla/joomla-cms/pull/11581 **Revert "Performance gain - new featured article"**
has fixed: https://github.com/joomla/joomla-cms/issues/11103  **New articles are created with last possible "ordering"** 

This PR was supposed to also be considered for merging, to address the performance issue,
(see my previous comment) but eventually this did not happen, so J3.6.3 has the known JTable::reorder() performance issue

ggppdk committed:
Revert wrong commit during rebase

mbabker commented:
Reading through this conversation and patch again, I'm still not 100% comfortable with this.  It still seems as though it has been coded against default MySQL configurations.  I've seen a couple tests for PostgreSQL, where frankly I'm not as concerned about it working since it usually behaves closer to MySQL anyway.  I haven't seen anything in here at all for SQL Server (and yes it's still supported right now so that needs to be validated).

Two things I think need to happen:
1) SQL Server testing; this CANNOT be merged without testing on that platform as long as we make a claim to supporting it
2) How can this be improved to be better aware of environmental limits with the different platforms?  This is hardcoded against MySQL's default configurations, what happens in the case where smaller limits are used or one of the platforms doesn't match the expectations added here?

I'm really not trying to be difficult, even if others think my arguments are BS.  But we are making changes at the base level of the database integration layer and the changes need to be properly validated and flexible enough to deal with lower end configurations.  So far, I think we're OK for default MySQL and PostgreSQL configurations.

csthomas commented:
There is a very old article that describe similar solution to this PR at http://www.databasejournal.com/features/mssql/article.php/1460001/Complex-Updates-Using-the-Case-Statement.htm

ggppdk commented:
> 2) How can this be improved to be better aware of environmental limits with the different platforms? This is hardcoded against MySQL's default configurations, what happens in the case where smaller limits are used or one of the platforms doesn't match the expectations added here?

The query start breaking into multiple queries at 10,000 records, which is about 350KBytes 
because 10,000 * 30 (or much less characters) + some overhead <= 350 KBytes

**Max query string size for MySQL, default is: 1MByte**
-- would anyone reconfinguring this to a lower limit and having a website with 10,000+ articles in the same category ?! (i have not seen such environments)

**Max query string size for postgresql: 1 Gbyte**

@mbabker 

So, this is well below the default limits, but even if we change the limit to e.g. 5,000 or to 3,000 records per query the performance difference will be small / very small
- tell me and i will just can just change it now, no real performance impact to lower the default from 10,000 records to 3,000 records

csthomas commented:
The other idea could be use `row_number` described at http://stackoverflow.com/questions/3047789/how-to-enumerate-returned-rows-in-sql which is supported on mssql, and postresql from 8.4. For mysql there could be a workaround with @ variables. This is only idea.

ggppdk commented:
@csthomas 

You mean this PR (scroll to the bottom to see the query)
https://github.com/joomla/joomla-cms/pull/8563/files

mbabker commented:
I don't know what can be done, but all I know is this is coded against MySQL's defaults, and based on your comment I guess PostgreSQL has a bigger default.  Are we sure SQL Server has a similar configuration?  If so then it should be fine, but if not then this needs to be addressed.

Also, as I pointed out above, SQL Server has a stricter configuration with the number of records/values that can be set with queries.  Maybe this isn't an issue with UPDATE queries as being issued here, but when I wrote the SQL Server adapter for Smart Search it was a very real issue with INSERT queries and the 10,000 records limit you have here exceeds the 1,000 record limitation I had to work around.  So even if the configurations are OK for the query size (in terms of bytes), we may still have an issue with this aspect of it.  Unless I scrolled past something, that was another thing I don't remember seeing an answer for.

Again, so far it seems like we're doing fine for MySQL and PostgreSQL for the most part.  But there's still one environment with no testing that we claim to support.  At least we're dropping it at 4.0 but we should still make sure it's working as long as 3.x is supported.

ggppdk commented:
hhmm, 
- i will test against postgress (@alikon had tested)
- about SQL server, i will try to install sql server lite (or something) and try to test with it too

Also even a very low limit of 500 records will have about 90%-95% of the performance benefit that 10,000 records limit per query gives

so we can just use 1,000 as limit of records and thus be sure of SQL server too

csthomas commented:
@ggppdk I thought about something more cross platform like #12839.

ggppdk commented:
I ll find time to test

mbabker commented:
With the other PRs that have been merged addressing reordering performance, is this still valid?

ggppdk commented:
This PR is no longer needed
